### ISSUE TITLE
Type inference error with method generic 'this' typing

### ISSUE THREAD

**Issue by HerringtonDarkholme** [Microsoft/TypeScript#10468]
Sorry for the undescriptive title but the behavior is too strange for me to describe...

Related to https://github.com/Microsoft/TypeScript/issues/9673 and https://github.com/Microsoft/TypeScript/pull/9746 with subtle difference

Maybe it's still problem for instantiating `this`? @sandersn 

**TypeScript Version:**   2.0.0

**Code**

Declaration
```ts
// A *self-contained* demonstration of the problem follows...
declare class VueTyped<T> {
  data<D>(d: D): VueTyped<T & D>;
  method<M extends { [k:string]: (this: T) => ({} | void) }>(m: M): VueTyped<T & M>
  static init(): VueTyped<{}>
}
```

Usage: 

When under `--noImplicitThis` flag, this code reports `this` has `implicit any`.
No error when `implictThis` is allowed.

```ts
VueTyped.init()
  .data({msg: '123'})
  .method({
    method() {
	  this.msgs.length // error here but reported wrongly
	  return this.msg 
    }
  })
```

Another usage:
When return value isn't from `this`, compiler has a false alarm.

```ts
VueTyped.init()
  .data({msg: '123'})
  .method({
    method() {
      this.msg // error her but should not
      return '123'
    }
  })
```

**Expected behavior:**
For the first usage, `this.msgs` should trigger `Property not exist` error.
For the second usage, `this.msg` should not report any error.

**Actual behavior:** 

No implicit this errors for both.

**Comment by sandersn** [comment]
Shouldn't the second example trigger the error "'msg' does not exist"? As far as I can tell, `this: T` and `T = {}` based on the call to `VueTyped.init(): VueTyped<{}>`

**Comment by sandersn** [comment]
Never mind, I didn't see that the definition of `data: VueTyped<T & D>`. 

Without `--noImplicitThis`, I don't get an error for either case. With `--noImplicitThis`, I get two "this is implicitly any" errors. That's because there is no contextual this-typing happening at all. I think it's because methods aren't aware of the signature that should be coming from the string index.

**Comment by sandersn** [comment]
Here's a simplified repro. I'm going to see if #9746 fixes this.

```ts
declare function infer<T, M extends { [k:string]: (this: T, hunch: T) => ({} | void) }>(t: T, m: M): void;
infer({msg: '123'}, {
  method() {
    this.msg // error her but should not
    return '123'
  }
})
infer({msg: '123'}, {
  method(hunch) {
    this.msgs.length // error here but reported wrongly
    return this.msg // shouldn't error here either
  }
})
```

**Comment by sandersn** [comment]
Yes, this is fixed by #9746

**Comment by HerringtonDarkholme** [comment]
Thanks! Going to try this in nightly. TypeScript is just so awesome!
